File: DACMASDK/DACMASDK-Sample-ObjectiveC/DACSDKMA-Sample-ObjectiveC/ViewController.m
//
//  ViewController.m
//  DACSDKMA-SamplePlayer-ObjectiveC
//
//  Copyright © 2016 D.A.Consortium Inc. All rights reserved.
//

#import <AVFoundation/AVFoundation.h>
#import <UIKit/UIKit.h>
#import <DACSDKMA/DACSDKMA-Swift.h>
#import "DACSDKMA_Sample_ObjectiveC-Swift.h"

#import "ViewController.h"

// サンプル動画URL
NSString *const contentUrl = @"http://vjs.zencdn.net/v/oceans.mp4";

// サンプルアドタグURL
NSString *const adTagUri = @"https://saxp.zedo.com/asw/fnsr.vast?n=2696&c=25/11&d=17&s=2&v=vast2&pu=__page-url__&ru=__referrer__&pw=__player-width__&ph=__player-height__&z=__random-number__";


@interface ViewController() <DACSDKMAAdsLoaderDelegate, DACSDKMAAdsManagerDelegate>

// 動画ビュー
@property(nonatomic, strong) UIView *videoView;
// 動画コンテンツプレイヤー
@property(nonatomic, strong) AVPlayer *contentPlayer;
// 動画再生ボタン
@property(nonatomic, strong) UIButton *playButton;

// SDK用変数
@property(nonatomic, strong) DACSDKMAAdsLoader *dacAdsLoader;
@property(nonatomic, strong) DACSDKMAAdsManager *dacAdsManager;
@property(nonatomic, strong) DACSDKMAAdDefaultController *dacAdController;

@end


@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // videoViewをセットする
    self.videoView = [[UIView alloc] initWithFrame: CGRectMake(20, 20, self.view.bounds.size.width - 40, self.view.bounds.size.height - 40)];
    self.videoView.backgroundColor = [UIColor colorWithWhite:0.0 alpha: 0.5];
    self.videoView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    [self.view addSubview:self.videoView];
    
    // playButtonをセットする
    self.playButton = [[UIButton alloc] initWithFrame:self.videoView.bounds];
    [self.playButton setTitle:@"▶️" forState:UIControlStateNormal];
    [self.playButton addTarget:self action:@selector(onPlayButtonTouch:) forControlEvents:UIControlEventTouchUpInside];
    self.playButton.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    self.playButton.layer.zPosition = MAXFLOAT;
    [self.videoView addSubview:self.playButton];
    
    // 動画コンテンツプレイヤーを使うための準備をする。
    [self setUpContentPlayer];
    
    // SDKを使うための準備をする
    [self setupAdsLoader];
}

- (IBAction)onPlayButtonTouch:(id)sender {
    self.playButton.hidden = YES;
    [self requestAds];
}


- (void)setUpContentPlayer {
    self.contentPlayer = [AVPlayer playerWithURL:[NSURL URLWithString:contentUrl]];
    AVPlayerLayer *playerLayer = [AVPlayerLayer playerLayerWithPlayer:self.contentPlayer];
    playerLayer.frame = self.videoView.layer.bounds;
    [self.videoView.layer addSublayer:playerLayer];
}

#pragma mark SDK Setup
- (void)setupAdsLoader {
    DACSDKMASettings *settings = [[DACSDKMASettings alloc] init];
    self.dacAdsLoader = [[DACSDKMAAdsLoader alloc] initWithSettings:settings];
    self.dacAdsLoader.delegate = self;
}

- (void)requestAds {
    DACSDKMAAdContainer *adContainer = [[DACSDKMAAdContainer alloc] initWithView:self.videoView companionSlots:nil];
    DACSDKMAAdsRequest *request = [[DACSDKMAAdsRequest alloc] initWithAdTagURI:adTagUri adContainer: adContainer contentPlayhead: nil];
    [self.dacAdsLoader requestAds:request];
}


#pragma mark AdsLoader Delegates
// 広告データの読み込みが正常に完了した際に呼ばれます。
- (void)dacSdkMaAdsLoader:(DACSDKMAAdsLoader *)loader adsLoadedWithData:(DACSDKMAAdsLoadedData *)data {
    self.dacAdsManager = data.adsManager;
    self.dacAdsManager.delegate = self;
    [self.dacAdsManager load];
    
    self.dacAdController = [[DACSDKMAAdDefaultController alloc] initWithAdsManager:self.dacAdsManager];
}

// 広告データの読み込みが失敗した際に呼ばれます。
- (void)dacSdkMaAdsLoader:(DACSDKMAAdsLoader *)loader failedWithErrorData:(DACSDKMAAdLoadingErrorData *)adErrorData {
    NSLog(@"Error: dacSdkMaAdsLoader:(loader) failedWithErrorData(adErrorData) = %@", adErrorData.adError.message);
    [self.contentPlayer play];
}

#pragma mark AdsManager Delegates
// DACSDKMAAdEventが発生した際に呼ばれます。
- (void)dacSdkAdsManager:(DACSDKMAAdsManager *)adsManager didReceiveAdEvent:(DACSDKMAAdEvent *)adEvent {
    if (adEvent.type == DACSDKMAAdEventTypeDidLoad) {
        [self.dacAdsManager play];
    }
    else if (adEvent.type == DACSDKMAAdEventTypeDidAllAdsComplete) {
        [self.dacAdsManager clean];
        [self.dacAdController clean];
    }
}

// DACSDKMAAdErrorが発生した際に呼ばれます。
- (void)dacSdkAdsManager:(DACSDKMAAdsManager *)adsManager didReceiveAdError:(DACSDKMAAdError *)adError {
    NSLog(@"Error: dacSdkAdsManager:(adsManager) didReceiveAdError:(adError) = %@", adError.message);
}

// 動画広告が再生開始、レジュームした際に呼ばれます。アプリのビデオコンテンツに停止を要求します。
- (void)dacSdkAdsManagerDidRequestContentPause:(DACSDKMAAdsManager *)adsManager {
    [self.contentPlayer pause];
}

// 動画広告が一時停止、正常終了・エラー終了した際に呼ばれます。アプリのビデオコンテンツに再生を要求します。
- (void)dacSdkAdsManagerDidRequestContentResume:(DACSDKMAAdsManager *)adsManager {
    [self.contentPlayer play];
}

@end

File: DACMASDK-Scroll-Sample-Swift/DACMASDK-Scroll-Sample-Swift/ViewController.swift
//
//  ViewController.swift
//  DACSDKMA-SampleScrollView-Swift
//
//  Copyright © 2016 D.A.Consortium Inc. All rights reserved.
//

import UIKit
import DACSDKMA

class ViewController: UIViewController, DACSDKMAAdsLoaderDelegate, DACSDKMAAdsManagerDelegate {

    // リクエスト先アドタグURI
    let adTagUri = "https://saxp.zedo.com/asw/fnsr.vast?n=2696&c=25/11&d=17&s=2&v=vast2&pu=__page-url__&ru=__referrer__&pw=__player-width__&ph=__player-height__&z=__random-number__"

    var scrollView: UIScrollView!
    
    // 広告掲載するView
    var adView: UIView!
    
    // SDKで使用する変数を宣言する
    var dacSdkAdsMaSettings: DACSDKMASettings?   = nil
    var dacAdsLoader: DACSDKMAAdsLoader?  = nil
    var dacAdsManager: DACSDKMAAdsManager? = nil
    var dacAdController: DACSDKMAAdDefaultController? = nil
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // self.viewのサイズを取得する
        let selfViewCGRect:     CGRect  = self.view.bounds
        let selfViewWidth:      CGFloat = selfViewCGRect.width
        let selfViewHeight:     CGFloat = selfViewCGRect.height
        
        // scrollViewを設定する
        self.scrollView = UIScrollView(frame: CGRectMake(0, 0, selfViewWidth, selfViewHeight))
        self.scrollView.backgroundColor  = UIColor.blueColor()
        self.scrollView.contentSize      = CGSizeMake(selfViewWidth, selfViewHeight*3)
        
        // adViewを設定する
        self.adView = UIView(frame: CGRectMake(0, selfViewHeight-100, selfViewWidth, 300))
        self.adView.backgroundColor = UIColor.greenColor()
        self.scrollView.addSubview(adView)
        
        // self.viewにscrollViewをaddSubviewする
        self.view.addSubview(scrollView)
        
        self.requestAds()
    }
    
    // VAST広告をリクエストする
    func requestAds() {
        self.dacSdkAdsMaSettings = DACSDKMASettings()
        self.dacAdsLoader =  DACSDKMAAdsLoader(settings: self.dacSdkAdsMaSettings!)
        self.dacAdsLoader?.delegate = self
        
        //広告動画を掲載するViewとリクエスト先アドサーバのURLをリクエストにセットする
        let adContainer: DACSDKMAAdContainer = DACSDKMAAdContainer(view: self.adView!)
        let request: DACSDKMAAdsRequest = DACSDKMAAdsRequest(adTagURI: self.adTagUri, adContainer: adContainer)
        self.dacAdsLoader?.requestAds(request)
    }
    
    // VASTレスポンスの処理が終わった時に実行される
    func dacSdkMaAdsLoader(loader: DACSDKMAAdsLoader, adsLoadedWithData data: DACSDKMAAdsLoadedData) {
        self.dacAdsManager = data.adsManager
        self.dacAdsManager?.delegate = self
        // 広告動画ファイルのダウンロードする
        self.dacAdsManager?.load()
        
        self.dacAdController = DACSDKMAAdDefaultController(adsManager: data.adsManager)
    }
    
    // VASTリクエスト/レスポンスでエラーが呼ばれたときに実行される
    func dacSdkMaAdsLoader(loader: DACSDKMAAdsLoader, failedWithErrorData adErrorData: DACSDKMAAdLoadingErrorData) {
        // エラーログを出力する。
        NSLog("dacSdkMaAdsLoader: failedWithErrorData: \(adErrorData.adError.message)")
    }
    
    // AdsManagerのイベントが発生した時に実行される
    func dacSdkAdsManager(adsManager: DACSDKMAAdsManager, didReceiveAdEvent adEvent: DACSDKMAAdEvent) {
        if adEvent.type == DACSDKMAAdEventType.DidLoad {
            // 広告動画ファイルがダウンロードされたら、広告動画を再生する
            adsManager.play()
        }
        else if adEvent.type == DACSDKMAAdEventType.DidAllAdsComplete{
            adsManager.clean()
            self.dacAdsManager = nil
            self.dacAdController?.clean()
            self.dacAdController = nil
        }
    }
    
    // AdsManagerにエラーが発生した時に実行される
    func dacSdkAdsManager(adsManager: DACSDKMAAdsManager, didReceiveAdError adError: DACSDKMAAdError) {
        // エラーログを出力し、本編動画を再生する
        NSLog("dacSdkMaAdsManager: didReceiveAdError: \(adError.message)")
    }
    
    func dacSdkAdsManagerDidRequestContentPause(adsManager: DACSDKMAAdsManager) {
        // アプリに動画プレーヤーがないので何もしない
    }
    
    func dacSdkAdsManagerDidRequestContentResume(adsManager: DACSDKMAAdsManager) {
        // アプリに動画プレーヤーがないので何もしない
    }
}

